platforms/chibios/drivers/i2c_master.c

```diff
@@ -111,11 +111,33 @@ i2c_status_t i2c_writeReg(uint8_t devaddr, uint8_t regaddr, const uint8_t* data,
     return chibios_to_qmk(&status);
 }
 
+i2c_status_t i2c_writeReg16(uint8_t devaddr, uint16_t regaddr, const uint8_t* data, uint16_t length, uint16_t timeout) {
+    i2c_address = devaddr;
+    i2cStart(&I2C_DRIVER, &i2cconfig);
+
+    uint8_t complete_packet[length + 2];
+    for (uint8_t i = 0; i < length; i++) {
+        complete_packet[i + 2] = data[i];
+    }
+    complete_packet[0] = regaddr >> 8;
+    complete_packet[1] = regaddr & 0xFF;
+
+    msg_t status = i2cMasterTransmitTimeout(&I2C_DRIVER, (i2c_address >> 1), complete_packet, length + 2, 0, 0, TIME_MS2I(timeout));
+    return chibios_to_qmk(&status);
+}
+
 i2c_status_t i2c_readReg(uint8_t devaddr, uint8_t regaddr, uint8_t* data, uint16_t length, uint16_t timeout) {
     i2c_address = devaddr;
     i2cStart(&I2C_DRIVER, &i2cconfig);
     msg_t status = i2cMasterTransmitTimeout(&I2C_DRIVER, (i2c_address >> 1), &regaddr, 1, data, length, TIME_MS2I(timeout));
     return chibios_to_qmk(&status);
 }
 
+i2c_status_t i2c_readReg16(uint8_t devaddr, uint16_t regaddr, uint8_t* data, uint16_t length, uint16_t timeout) {
+    i2c_address = devaddr;
+    i2cStart(&I2C_DRIVER, &i2cconfig);
+    msg_t status = i2cMasterTransmitTimeout(&I2C_DRIVER, (i2c_address >> 1), &regaddr, 2, data, length, TIME_MS2I(timeout));
```

Comment:
This depends on the endianness of the cpu, and also I think ARM is little-endian by default, so I think this is using the opposite endiannes of the register address, as compared to the write16 function.

By the way, that brings up another topic: Are all 16-bit register i2c devices generally big endian? If not we may need two variants of these functions.

Reply:
Both AVR and ARM are little endian, so I don't think there's an issue here?

Reply:
There is an issue. The register address is stored little-endian in memory, but needs to be sent in big-endian order. All other functions do big-endian order, so I'm assuming that's the correct i2c endianness. Hopefully all i2c 16-bit chips use big-endian communication. It might be worth specifying in the functions documentation that the register address is sent big-endian. If you just call chibios's i2cMasterTransmitTimeout function with the pointer to uint16_t it will transmit little-endian order, so reading register 0x0001 and then writing it back will write register 0x0100.

Reply:
Okay, this latest commit (ignoring the formatting) should resolve the ordering part at least. I don't know if all I2C devices communicate in big endian, but it would make sense, I think.